### PR TITLE
Add rest api for realtime segment size tuning.

### PR DESCRIPTION
**Pull request by mayankshriv**
## Description
Added a controller api to suggest segment size for realtime tables.

* Api call looks like: `http://localhost:9000/tables/autotune/airlineStats?type=realtime`
* The high level idea of the algorithm is to allow a certain percentage of heap usage (0.3x of Xmx) at peak time. At peak time, we would have max Ps consuming segments, and Pm parallel segment builds. This is then used to derive segment size that would keep heap usage within desired limit. More details documented in the code. 
* The output of API is a JSON record with:
   - New table config containing the `realtime.segment.flush.threshold.segment.size` config set.
   - Suggested level of max-parallel segment builds. The algorithm also tries to reduce this value to limit heap usage.
* The table config returned uses the new `StreamIngestionConfig` and `BatchIngestionConfig`, instead of the deprecated `StreamConfig`.

TODO:
*  Add `apply` param to the api to be able to apply the suggestions. This is a simply change, but the PR intentionally shies away from making this change, in order to first collect feedback on the algorithm.
* Currently segment size proposed is within a min (10MB) and a max (100MB) range. Make these limits configurable.
* Refine the algorithm as needed.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableAutoTunerResource.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.common.utils.config.TableConfigUtils.updateTableConfigWithSegmentSize;
+
+
+/**
+ * Rest apis for auto-tuning tables.
+ */
+@Api(tags = Constants.CLUSTER_TAG)
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableAutoTunerResource.java:63]: Shall this be under the `TABLE_TAG`?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableAutoTunerResource.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+@Path("/autotune/tables")
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableAutoTunerResource.java:64]: Suggest making it `/tables/autotune`
We have similar apis such as `/tables/recommender`

Comment (mayankshriv) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableAutoTunerResource.java:64]: Hmm, I went by `/debug/table/`, where the idea was to have `/debug/cluster` or other entities. I guess tuning only applies to table? In which case, we can go with your suggestion, wdyt?

Comment (mayankshriv) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableAutoTunerResource.java]: On second thought, took your suggestion.

**Comment by siddharthteotia** [comment]
We recently added https://github.com/apache/incubator-pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java
in the recommendation engine. This is also invokable via Rest API . Can we enhance that instead of adding a new API ?
cc @sajjad-moradi @mcvsubbu

**Comment by siddharthteotia** [comment]
`New table config containing the realtime.segment.flush.threshold.segment.size config set.`

https://github.com/apache/incubator-pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java has support for this so I suggest enhancing that.

**Comment by mayankshriv** [comment]
@siddharthteotia @sajjad-moradi  Does the existing feature require a lot of inputs from the user? We have seen that does not work in the OSS because it is chicken/egg problem. In this PR, there is no input needed from the user. We can separate the api from the algorithm, I'd like to keep this algorithm as it does not require any inputs from the user. As for api, can you guys suggest the exact change in the api that you would like?

**Comment by mayankshriv** [comment]
This needs more discussion, closing the PR for now.

**Pull request closed without merging**